qiskit/circuit/controlflow/for_loop.py

```diff
@@ -19,12 +19,75 @@
 
 from qiskit.circuit.parameter import Parameter
 from qiskit.circuit.exceptions import CircuitError
+
 from .control_flow import ControlFlowOp
+from qiskit.circuit.classical.expr import Var
+from qiskit.circuit.classical.types import Uint
+
 
 if TYPE_CHECKING:
     from qiskit.circuit import QuantumCircuit
 
 
+class DynamicRange:
+    """
+    A class representing a dynamic range of integers.
+    This class is used to represent a range of integers that can be
+    defined at runtime. It is used in the context of quantum circuits
+    that may require dynamic ranges for control flow operations.
+    The class is initialized with a starting value, an optional stopping
+    value, and a step size. The starting value can be an integer or a
+    variable, and the stopping value can also be an integer or a variable.
+    The step size is an integer or a variable that defines the increment
+    """
+
+    def __init__(self, start: int | Var, stop: Optional[int | Var] = None, step: int | Var = 1):
+        """
+        Args:
+            start: The starting value of the range.
+            stop: The stopping value of the range.
+            step: The step size for the range.
+        """
+        if stop is None:
+            start, stop = 0, start
+
+        for value, name in [(start, "start"), (stop, "stop"), (step, "step")]:
+            if not isinstance(value, (int, Var)):
+                raise CircuitError(
+                    f"DynamicRange {name} value must be an int or Var, but got {type(value)}."
+                )
```

Comment:
In general, it's best not to do this in the default constructor - in Python, the accepted types are documented in the method, and doing manual runtime type-checking slows down the happy paths for the vast majority of users, without being an adequate prevention of typing errors.  I know we do it a lot in Qiskit, but it's a pattern we're trying to wind down, because in practice while each individual one is "not much", the whole effect is a library that's sluggish.

Certainly the `expr.Expr` subclass tree treats the classes as the "fast-path constructors", and leaves all the type-checking for the type-inferring/lifting function calls (where it is fine to include).